src/oci/oci_runtime.rs

```diff
@@ -0,0 +1,734 @@
+//! Interface to [oci_runtime][0], a OCI compliant cli for spawning and running containers.
+//!
+//! [0]: https://github.com/opencontainers/oci_runtime
+
+#![allow(dead_code)]
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use derive_builder::Builder;
+use dyn_clone::{clone_trait_object, DynClone};
+use getset::{Getters, Setters};
+use std::{
+    fmt::{self, Debug},
+    path::{Path, PathBuf},
+    process::Output,
+    string::ToString,
+};
+use strum::AsRefStr;
+use tokio::process::Command;
+
+#[derive(Builder, Debug, Getters, Setters)]
+#[builder(pattern = "owned", setter(into))]
+// oci_runtime is the main structure to be used when interacting with the container runtime.
+pub struct OCIRuntime {
+    #[getset(get, set)]
+    #[builder(private, default = "Box::new(DefaultOCIRuntimeExecCommand)")]
+    /// The executor for the OCIRuntime
+    exec: Box<dyn ExecCommand>,
+
+    #[get]
+    /// Path to the oci_runtime binary
+    binary: PathBuf,
+}
+
+impl OCIRuntime {
+    /// Run OCIRuntime with the provided subcommand and args and return the output if the command execution succeeds.
+    /// This can still mean that oci_runtime itself failed, which can be verified via the exist status
+    /// of the output.
+    pub async fn run(&self, subcommand: &Subcommand, args: &[GlobalArgs]) -> Result<Output> {
+        self.exec()
+            .run_output(self.binary(), &subcommand.build_cmd()[..], args)
+            .await
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+/// Defaultoci_runtimeExecCommand is a wrapper which can be used to execute oci_runtime in a standard way.
+struct DefaultOCIRuntimeExecCommand;
+
+impl ExecCommand for DefaultOCIRuntimeExecCommand {}
+
+#[async_trait]
+trait ExecCommand: Debug + DynClone + Send + Sync {
+    /// Run a command and return its `Output`.
+    async fn run_output(
+        &self,
+        binary: &Path,
+        cmd: &[String],
+        global_args: &[GlobalArgs],
+    ) -> Result<Output> {
+        Command::new(binary)
+            .args(cmd)
+            .args(global_args.iter().map(ToString::to_string))
+            .output()
+            .await
+            .context("run OCIRuntime")
+    }
+}
+
+clone_trait_object!(ExecCommand);
+
+#[derive(AsRefStr, Clone, Debug, Hash, Eq, PartialEq)]
+#[strum(serialize_all = "kebab_case")]
+pub enum Subcommand {
+    /// Checkpoint a running container
+    Checkpoint((String, Vec<CheckpointArgs>)),
```

Comment:
If the first value is always the `container_id`, then I'd suggest to make it a type alias:

```rust
type ContainerId = String;
```